test/puppetlabs/puppetdb/integration/fixtures.clj

```diff
@@ -0,0 +1,375 @@
+
+(defrecord PuppetDBTestServer [-info-map app]
+  TestServer
+  (info-map [_] -info-map)
+
+  java.lang.AutoCloseable
+  (close [_] (tk-app/stop app)))
+
+(defn stringify-keys-with-ns [m]
+  (let [f (fn [[k v]] (if (keyword? k) [(subs (str k) 1) v] [k v]))]
+    ;; only apply to maps
+    (walk/postwalk (fn [x] (if (map? x) (into {} (map f x)) x)) m)))
+
+;; TODO: Get this fixed version into the clj-typesafe-config library and use it
+;; from there
+(defn map->hocon-string
+  "Serialize the clojure data structure `m` to string using the
+  typesafe config format. `m` is typically the result of a
+  `config-file->map` or `reader-map` with some modifications."
+  [m]
+  (-> m
+      stringify-keys-with-ns
+      ConfigValueFactory/fromAnyRef
+      .render))
+
+(defn adjust-config-file [in-file out-file overrides]
+  (spit out-file (-> (hocon/config-file->map in-file)
+                     (ks/deep-merge overrides)
+                     map->hocon-string)))
+
+(defn start-test-puppetdb [bootstrap-config-file config-file config-overrides bind-attempts]
+  (when (zero? bind-attempts)
+    (throw (RuntimeException. "Repeated attempts to bind port failed, giving up")))
+  (let [port (svc-utils/open-port-num)
+        tmp-puppetdb-conf (str (tu/temp-file "puppetdb" ".conf"))
+        _ (adjust-config-file config-file tmp-puppetdb-conf
+                              (-> config-overrides
+                                  (assoc-in [:jetty :ssl-port] port)
+                                  (assoc-in [:global :vardir] (str (tu/temp-dir)))))
+        base-url {:protocol "https"
+                  :host "localhost"
+                  :port port
+                  :prefix "/pdb/query"
+                  :version :v4}]
+    (try
+      (swap! dispatch/metrics svc-utils/clear-counters!)
+      {:app (tu/without-jmx
+             (tkbs/parse-and-bootstrap bootstrap-config-file
+                                       {:config tmp-puppetdb-conf}))
+       :base-url base-url}
+
+      (catch java.net.BindException e
+        (log/errorf e "Error occured when Jetty tried to bind to port %s, attempt #%s"
+                    port bind-attempts)
+        (start-test-puppetdb bootstrap-config-file config-file config-overrides (dec bind-attempts))))))
+
+(defn run-puppetdb [postgres-server config-overrides]
+  (let [{:keys [app base-url]} (start-test-puppetdb "test-resources/integration-bootstrap.cfg"
+                                                    "test-resources/integration-puppetdb.conf"
+                                                    (ks/deep-merge config-overrides
+                                                                   {:database (info-map postgres-server)})
+                                                    10)
+        {:keys [host port]} base-url]
+    (PuppetDBTestServer. {:hostname host
+                          :port port
+                          :query-base-url (assoc base-url :prefix "/pdb/query" :version :v4)
```

Comment:
Any objections to switching this to use functions in services.clj? Or adapting those to allow a base-url?